lib/ansible/modules/remote_management/manageiq/manageiq_alerts.py

```diff
@@ -0,0 +1,363 @@
+        the resource as an object if it exists in manageiq, Fail otherwise.
+    """
+    resource = manageiq.find_collection_resource_by(collection_name, **params)
+    if resource:
+        return resource
+    else:
+        msg = "{collection_name} where {params} does not exist in manageiq".format(
+            collection_name=collection_name, params=str(params))
+    module.fail_json(msg=msg)
+
+
+class ManageIQAlert(object):
+    """ Represent a ManageIQ alert
+    """
+    def __init__(self, alert):
+        self.description = alert['description']
+        self.db = alert['db']
+        self.enabled = alert['enabled']
+        self.options = alert['options']
+        self.hash_expression = None
+        self.miq_expressipn = None
+
+        if 'hash_expression' in alert:
+            self.hash_expression = alert['hash_expression']
+        if 'miq_expression' in alert:
+            self.miq_expression = alert['miq_expression']
+            if 'exp' in self.miq_expression:
+                # miq_expression is a field that needs a special case, because
+                # it's returned surrounded by a dict named exp even though we don't
+                # send it with that dict.
+                self.miq_expression = self.miq_expression['exp']
+
+    def __eq__(self, other):
+        """ Compare two ManageIQAlert objects
+        """
+        return self.__dict__ == other.__dict__
+
+
+class ManageIQAlerts(object):
+    """ Object to execute alert management operations in manageiq.
+    """
+
+    def __init__(self, manageiq):
+        self.manageiq = manageiq
+
+        self.module = self.manageiq.module
+        self.api_url = self.manageiq.api_url
+        self.client = self.manageiq.client
+        self.url = '{api_url}/alert_definitions'.format(api_url=self.api_url)
+
+    def get_alerts(self):
+        """ Get all alerts from ManageIQ
+        """
+        try:
+            response = self.client.get(self.url + '?expand=resources')
+        except Exception as e:
+            self.module.fail_json(msg="Failed to query alerts: {error}".format(error=e))
+        return response.get('resources', [])
+
+    def validate_hash_expression(self, expression):
+        """ Validate a 'hash expression' alert definition
+        """
+        # hash expressions must have the following fields
+        for key in ['options', 'eval_method', 'mode']:
+            if key not in expression:
+                msg = "Hash expression is missing required field {key}".format(key=key)
+                self.module.fail_json(msg)
+
+    def create_alert_dict(self, params):
+        """ Create a dict representing an alert
+        """
+        if params['expression_type'] == 'hash':
+            # hash expression supports depends on https://github.com/ManageIQ/manageiq-api/pull/76
+            self.validate_hash_expression(params['expression'])
+            expression_type = 'hash_expression'
+        else:
+            expression_type = 'miq_expression'
```

Comment:
consider having miq_expression, hash_expression params.
- user would have to write just 1 param instead of currently needed 2 (type + expression)
- can natively document & validate each option differently
- this function mostly goes away :)  but this is least consideration.

Reply:
I tried that in one of the early iterations, and it made the code more complicated because it'd need more conditions and extra validation, for example - I don't see a way to use Ansible's require_if to require one of two options but never both - so more conditions would be needed.